### PR TITLE
Use Footbumpers to spawn unseen obstacles

### LINKED ISSUE #502: Use foot bumpers to detect obstacles

### PR DESCRIPTION
**Pull request by PatrickGoettsch**
## Introduced Changes

The robot makes use of the foot bumper to spawn obstacles it recognizes by stepping into it.

Fixes #502 , #510 

## ToDo / Known Issues

Needs to be tested in test games, to see potential issues

## Ideas for Next Iterations (Not This PR)
- Think about having different life times for sonar obstacles and footbumper obstacles
- Merge counter for center obstacles

## How to Test

Has to be enabled in config
Needs to be tested when you let a robot walk and observe in twix if it spawns no existing obstacles.
Tape one fottbumper to force constant triggering, and see if it does not spawn obstacles in that case.

### CODE REVIEW

File: crates/control/src/foot_bumper_filter.rs [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+
+        let left_point = point![
+            context.sensor_angle.cos() * *context.obstacle_distance,
+            context.sensor_angle.sin() * *context.obstacle_distance
+        ];
+        let right_point = point![
+            context.sensor_angle.cos() * *context.obstacle_distance,
+            -context.sensor_angle.sin() * *context.obstacle_distance
+        ];
+        let middle_point = point![*context.obstacle_distance, 0.0];
```
Comment (schluis) [inline crates/control/src/foot_bumper_filter.rs:185]: This is constant. Maybe just compute this once in `new`.

File: etc/parameters/default.json [hunk 2]
```diff
@@ -50,6 +50,16 @@
       "cc_optical_center": [0.5, 0.5]
     }
   },
+  "foot_bumper_filter": {
+    "activations_needed": 2,
+    "acceptance_duration": { "nanos": 0, "secs": 1 },
+    "buffer_size": 15,
+    "enabled": false,
+    "number_of_true_elements_in_buffer_for_defective_declaration": 13,
+    "number_of_true_elements_in_buffer_to_reset_in_use": 2,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply etc/parameters/default.json:59]: Done

File: etc/parameters/default.json [hunk 3]
```diff
@@ -50,6 +50,16 @@
...
+    "obstacle_distance": 0.02,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply etc/parameters/default.json:60]: Would leave this as a tuning parameter to be evaluated in games. 20cm seems to be okay.

File: crates/control/src/foot_bumper_filter.rs [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        "foot_bumper_filter.number_of_true_elements_in_buffer_for_defective_declaration",
+    >,
+    pub number_of_true_elements_in_buffer_to_reset_in_use:
+        Parameter<i32, "foot_bumper_filter.number_of_true_elements_in_buffer_to_reset_in_use">,
+    pub obstacle_distance: Parameter<f32, "foot_bumper_filter.obstacle_distance">,
+    pub sensor_angle: Parameter<f32, "foot_bumper_filter.sensor_angle">,
+
+    pub cycle_time: Input<CycleTime, "cycle_time">,
+    pub fall_state: Input<FallState, "fall_state">,
+    pub sensor_data: Input<SensorData, "sensor_data">,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:48]: Done

File: crates/control/src/foot_bumper_filter.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+impl FootBumperFilter {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            left_count: 0,
+            right_count: 0,
+            last_left_time: None,
+            last_right_time: None,
+            left_pressed_last_cycle: false,
+            right_pressed_last_cycle: false,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:65]: Done

File: crates/control/src/foot_bumper_filter.rs [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+#[context]
+pub struct CreationContext {}
+
+#[context]
+pub struct CycleContext {
+    pub foot_bumper_values: AdditionalOutput<FootBumperValues, "foot_bumper_values">,
+    pub acceptance_duration: Parameter<Duration, "foot_bumper_filter.acceptance_duration">,
+    pub activations_needed: Parameter<i32, "foot_bumper_filter.activations_needed">,
+    pub buffer_size: Parameter<i32, "foot_bumper_filter.buffer_size">,
+    pub enabled: Parameter<bool, "foot_bumper_filter.enabled">,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:36]: I see what you mean and thought, that it is clear from the context that the leading is foot_bumper_filter and enabled is than reflecting if it is enabled or not.

File: crates/control/src/foot_bumper_filter.rs [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                        self.left_count = 0;
+                        self.left_pressed_last_cycle = false;
+                    }
+                }
+                Err(e) => {
+                    eprintln!("Duration elapsed failed: {e:?}");
+                    self.last_left_time = None;
+                    self.left_count = 0;
+                    self.left_pressed_last_cycle = false;
+                }
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:118]: Done, although I do not know why your way is better, but okay.

File: crates/control/src/foot_bumper_filter.rs [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+                        self.right_count = 0;
+                        self.right_pressed_last_cycle = false;
+                    }
+                }
+                Err(e) => {
+                    eprintln!("Duration elapsed failed: {e:?}");
+                    self.last_right_time = None;
+                    self.right_count = 0;
+                    self.right_pressed_last_cycle = false;
+                }
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:135]: Done

File: crates/control/src/foot_bumper_filter.rs (hunk 1 above)
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:185]: Done

[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:185]: Done

File: etc/parameters/default.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (schluis) [reply etc/parameters/default.json:60]: Agree, but 0.02 m are 2 cm 😉

File: crates/control/src/foot_bumper_filter.rs [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        if !self.left_in_use
+            && left_count <= *context.number_of_true_elements_in_buffer_to_reset_in_use
+        {
+            self.left_in_use = true;
+        }
+        if !self.right_in_use
+            && right_count <= *context.number_of_true_elements_in_buffer_to_reset_in_use
+        {
+            self.right_in_use = true;
+        }
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:175]: Done

File: crates/control/src/foot_bumper_filter.rs [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    left_detection_buffer: VecDeque<bool>,
+    right_detection_buffer: VecDeque<bool>,
+}
+
+#[context]
+pub struct CreationContext {}
+
+#[context]
+pub struct CycleContext {
+    pub foot_bumper_values: AdditionalOutput<FootBumperValues, "foot_bumper_values">,
```
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs:32]: As I got told, this is currently a missing feature in our framework.

File: etc/parameters/default.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply etc/parameters/default.json:60]: ok

[the remark this answers is not in the document]
Comment (PatrickGoettsch) [reply etc/parameters/default.json:60]: Done

**Comment by PatrickGoettsch** [comment]
@schluis All requested changes are addressed

File: crates/control/src/foot_bumper_filter.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/foot_bumper_filter.rs]: 😆 forgot that, sorry

File: crates/control/src/foot_bumper_filter.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (schluis) [reply crates/control/src/foot_bumper_filter.rs:41]: I suggested that to reduce the number of parameters in our config. Is there a use case, where `"enabled": false,` and `"use_foot_bumper_measurements": true,`?

File: crates/control/src/foot_bumper_filter.rs [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        let fall_state = context.fall_state;
+
+        if !context.enabled {
+            return Ok(MainOutputs::default());
+        }
+
+        if context.sensor_data.touch_sensors.left_foot_left
```
Comment (schluis) [inline crates/control/src/foot_bumper_filter.rs:94]: Maybe extract `context.sensor_data.touch_sensors` in a local variable `touch_sensors`

**Review by schluis — CHANGES_REQUESTED** [review]

File: crates/control/src/foot_bumper_filter.rs (hunk 11 above)
Comment (PatrickGoettsch) [reply crates/control/src/foot_bumper_filter.rs]: Done

**Review by schluis — APPROVED** [review]
Nice!

**Pull request merged**